### PR TITLE
split multiple types into anyOf subschema

### PR DESCRIPTION
**Pull request by 6293**
Closes https://github.com/getsentry/json-schema-diff/issues/16

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -175,6 +176,20 @@ pub enum JsonSchemaType {
     Null,
 }
 
+impl From<JsonSchemaType> for InstanceType {
```
[the remark this answers is not in the document]
Comment (6293) [reply src/types.rs:179]: I like the latter idea

File: src/diff_walker.rs [hunk 2]
```diff
@@ -319,27 +339,95 @@ impl DiffWalker {
...
+            InstanceType::Object => ret.object = schema_object.object.clone(),
+            InstanceType::Array => ret.array = schema_object.array.clone(),
+            _ => (),
+        }
+        Schema::Object(ret)
+    }
+
+    /// Split a schema into multiple schemas, one for each type in the multiple type.
+    /// Returns the new schema and whether the schema was changed.
+    fn split_types(schema_object: &mut SchemaObject) -> (SchemaObject, bool) {
```
[the remark this answers is not in the document]
Comment (6293) [reply src/diff_walker.rs]: thank you for the insight

**Review by untitaker — APPROVED** [review]

**Comment by untitaker** [comment]
Thank you!

**Comment by 6293** [comment]
> This is a completely unrealistic schema (I hope), but I'm just curious, do you think it is possible to handle this case and produce no diff?

We should not convert them into `rhs`-ish style like the following:

```json
{"properties": {"foo": {"anyOf": [{"type": "integer"}, {"type": "string"}]}}}
```
 
Because converting `lhs` into this style is not always possible. Consider more complex example:

```json
"anyOf": [
    {"properties": {"foo": {"type": "integer"}, "bar": {"type": "integer"}}},
    {"properties": {"foo": {"type": "integer"}, "bar": {"type": "string"}}},
    {"properties": {"foo": {"type": "string"}, "bar": {"type": "integer"}}},
]
```

In this case, `foo` can be string only if `bar` is string. So we are not able to write like 

```json
"anyOf": [
    {"properties": {"foo": {"anyOf": [???]}, "bar": {"anyOf": [???]}}}
 ]
```

**Comment by 6293** [comment]
In `lhs`-ish style conversion, 

Better conversion would be from `rhs` to `lhs`. e.g.

```json
"anyOf": [{
    "properties": {
        "foo": {
            "type": ["integer", "string"]
        },
        "bar": {
            "type": ["integer", "string"]
        }
    }
}]
```

to

```json
"anyOf": [
    {"properties": {"foo": {"type": "integer"}, "bar": {"type": "integer"}}},
    {"properties": {"foo": {"type": "integer"}, "bar": {"type": "string"}}},
    {"properties": {"foo": {"type": "string"}, "bar": {"type": "integer"}}},
    {"properties": {"foo": {"type": "string"}, "bar": {"type": "string"}}},
]
```

**Comment by untitaker** [comment]
a fun problem to think about :) i am not sure if it is worth solving. if we go with the conversion from rhs to lhs as per your last comment, we should have some sort of limit in place to prevent combinatorial explosion and OOM. some good performance tests would be in https://github.com/getsentry/sentry-kafka-schemas/

**Pull request merged**